File: py/src/domain/services/transfer_service.py
import uuid
from src.domain.entities.wallet import Wallet
from src.domain.entities.transaction import Transaction
from src.domain.value_objects.money import Money

class TransferService:
    def transfer(self, source: Wallet, target: Wallet, amount: Money) -> Transaction:
        if source.id == target.id:
            raise ValueError("Source and target wallets must be different.")

        # Debit source
        source.debit(amount)

        # Credit target
        target.credit(amount)

        # Create and return transaction
        transaction_id = f"tx_{uuid.uuid4().hex[:9]}"
        return Transaction(transaction_id, source.id, target.id, amount)

File: js/src/domain/services/TransferService.ts
import { Wallet } from "../entities/Wallet";
import { Money } from "../value-objects/Money";
import { Transaction } from "../entities/Transaction";

export class TransferService {
  public transfer(source: Wallet, target: Wallet, amount: Money): Transaction {
    if (source.id === target.id) {
      throw new Error("Source and target wallets must be different.");
    }

    // Debit source
    source.debit(amount);

    // Credit target
    target.credit(amount);

    // Create and return transaction
    const transactionId = `tx_${Math.random().toString(36).substring(2, 11)}`;
    return new Transaction(transactionId, source.id, target.id, amount);
  }
}
